src/icrar/leap-accelerate/algorithm/casa/PhaseRotate.cc

```diff
@@ -99,20 +131,43 @@ namespace casalib
                 {
                     return std::arg(c);
                 };
+
+                if(!metadata.avg_data.is_initialized())
+                {
+                    throw icrar::exception("avg_data must be initialized", __FILE__, __LINE__);
+                }
+
                 casacore::Matrix<Radians> avg_data = MapCollection(metadata.avg_data.get(), getAngle);
-                casacore::Array<double> cal1 = icrar::casalib::multiply(metadata.Ad1, avg_data.column(0));// TODO: (IPosition(0, metadata.I1)); //diagonal???
-                casacore::Matrix<double> dInt = avg_data(Slice(0, 0), Slice(metadata.I.shape()[0], metadata.I.shape()[1]));
-                
+
+                auto indexes = ConvertVector(metadata.I1);
+
+                auto avg_data_t = ConvertMatrix(static_cast<Eigen::MatrixXd>(ConvertMatrix(avg_data)(indexes, 0))); // 1st pol only
+                casacore::Matrix<double> cal1 = icrar::casalib::multiply(metadata.Ad1, avg_data_t);
+                assert(cal1.shape()[1] == 1);
+
+                casacore::Matrix<double> dInt = casacore::Matrix<double>(metadata.I.size(), avg_data.shape()[1]);
+                dInt = 0;
+
                 for(int n = 0; n < metadata.I.size(); ++n)
                 {
-                    dInt[n] = avg_data(metadata.I) - metadata.A(IPosition(1, n)) * cal1;
+                    Eigen::VectorXi e_i = ConvertVector(metadata.I);
+                    Eigen::MatrixXd e_avg_data_slice = ConvertMatrix(avg_data)(e_i, Eigen::all);
+                    casacore::Matrix<double> avg_data_slice = ConvertMatrix(e_avg_data_slice);
```

Comment:
These three statements seem to be loop-invariant, can they be moved outside the `for` loop?

Reply:
correct, I recall doing this in a later pull request but it should be fine to do that change in this PR